File: data-pipeline/scripts/fetch_bills.py
"""
Fetch recently voted-on U.S. Congressional bills and merge them into the
current Congress's manifest.

Reads CONGRESS_API_KEY from the environment, queries the Congress.gov v3 API
for the current Congress, keeps bills whose latest passage-type action falls
within the last RECENT_DAYS, enriches each one, then merges the result into
docs/data/congress{NNN}_bills.json and rewrites docs/data/congresses.json.
"""

from __future__ import annotations

import argparse
import os
import sys
from collections import Counter
from datetime import datetime, timedelta, timezone
from typing import Any, Iterable

from _common import (
    CongressClient,
    ErrorCollector,
    LIST_PAGE_LIMIT,
    build_bill_records_parallel,
    current_congress,
    evaluate_bill,
    load_manifest,
    load_vote_refs,
    merge_records,
    rebuild_index,
    save_manifest,
)
from _votes import attach_vote_refs, strip_vote_refs

RECENT_DAYS = 60
LIST_PAGES_MAX = 8
SAMPLE_REJECTIONS = 8


def list_recent_bills(
    client: CongressClient,
    congress: int,
    cutoff: datetime,
) -> Iterable[dict[str, Any]]:
    """Yield bill summaries from /bill/{congress} sorted by updateDate desc."""
    from_dt = cutoff.strftime("%Y-%m-%dT%H:%M:%SZ")
    for page in range(LIST_PAGES_MAX):
        offset = page * LIST_PAGE_LIMIT
        body = client.get(
            f"/bill/{congress}",
            limit=LIST_PAGE_LIMIT,
            offset=offset,
            sort="updateDate desc",
            fromDateTime=from_dt,
        )
        bills = body.get("bills") or []
        if not bills:
            return
        yield from bills
        if len(bills) < LIST_PAGE_LIMIT:
            return


def _parse_args(argv: list[str] | None) -> argparse.Namespace:
    """Parse CLI flags, mirroring fetch_votes.py's argparse pattern.

    ``argv`` is normalised to ``[]`` when None so a bare ``main()`` call
    (as the existing integration tests make) parses to defaults instead of
    reading the ambient ``sys.argv`` — the real CLI entry point passes
    ``sys.argv[1:]`` explicitly from ``__main__`` below.
    """
    parser = argparse.ArgumentParser(
        description="Fetch recently voted-on bills and merge them into the "
                    "current Congress's manifest.",
    )
    parser.add_argument(
        "--congress", type=int, default=None,
        help="Congress number (default: the current Congress).",
    )
    parser.add_argument(
        "--recent-days", type=int, default=RECENT_DAYS,
        help="Re-fetch window in days: keep bills whose latest passage-type "
             f"action falls within this many days (default {RECENT_DAYS}). "
             "The run is bounded further by the fetcher's 8-page list cap. "
             "Must be between 1 and 365 — the ceiling keeps a manual refresh "
             "from ever becoming an unbounded backfill.",
    )
    args = parser.parse_args(argv if argv is not None else [])
    if not 1 <= args.recent_days <= 365:
        parser.error("--recent-days must be between 1 and 365")
    return args


def main(argv: list[str] | None = None) -> int:
    args = _parse_args(argv)

    api_key = os.environ.get("CONGRESS_API_KEY")
    if not api_key:
        print("CONGRESS_API_KEY is not set in the environment.", file=sys.stderr)
        return 2

    now = datetime.now(timezone.utc)
    cutoff = now - timedelta(days=args.recent_days)
    congress = args.congress if args.congress is not None else current_congress(now)
    print(f"Fetching bills for the {congress}th Congress (cutoff {cutoff.date()})")

    client = CongressClient(api_key)
    reject_counts: Counter[str] = Counter()
    rejection_samples: list[str] = []
    errors = ErrorCollector()
    kept_summaries: list[tuple[dict[str, Any], str]] = []
    total_evaluated = 0

    # Phase 1: paginated list + filter. Sequential because pagination is
    # stateful and the per-summary work (``evaluate_bill``) is pure-Python.
    for summary in list_recent_bills(client, congress, cutoff):
        total_evaluated += 1
        outcome, reason = evaluate_bill(summary, cutoff)
        if outcome is None:
            reject_counts[reason] += 1
            if reason == "no_outcome_match" and len(rejection_samples) < SAMPLE_REJECTIONS:
                action_text = (summary.get("latestAction") or {}).get("text") or ""
                ref = f"{summary.get('type')}{summary.get('number')}"
                rejection_samples.append(f"{ref}: {action_text[:140]}")
            continue
        kept_summaries.append((summary, outcome))

    # Phase 2: per-bill enrichment, fanned out across a thread pool. Each
    # build issues 3 sequential Congress.gov GETs; with N workers we get
    # roughly an N× speedup until the API's per-key rate limit binds.
    fresh_records, build_failures = build_bill_records_parallel(
        client, congress, kept_summaries, errors
    )
    reject_counts["build_error"] += build_failures

    # Drop duplicates after the parallel pass (rare; same bill could appear
    # twice across paginated list responses if its updateDate shifts).
    seen_ids: set[str] = set()
    deduped: list[dict[str, Any]] = []
    for record in fresh_records:
        if record["id"] in seen_ids:
            reject_counts["duplicate"] += 1
            continue
        seen_ids.add(record["id"])
        deduped.append(record)
    fresh_records = deduped
    # Same date-desc-then-id ordering as merge_records, so the "+ id" log
    # lines below are deterministic across runs and implementations.
    fresh_records.sort(key=lambda r: r["id"])
    fresh_records.sort(key=lambda r: r["latest_action"]["date"], reverse=True)

    for record in fresh_records:
        print(f"  + {record['id']}: {record['outcome']} on {record['latest_action']['date']}")

    print()
    print(f"Evaluated {total_evaluated} bills, kept {len(fresh_records)}.")
    if reject_counts:
        print("Rejections:")
        for reason, count in reject_counts.most_common():
            print(f"  - {reason}: {count}")
    if rejection_samples:
        print("Sample latestAction texts that did not match any outcome rule:")
        for sample in rejection_samples:
            print(f"  · {sample}")
    errors.print_summary(label="fetch_bills")

    existing = load_manifest(congress)
    # ``votes`` is derived data: strip it before merging so equality compares
    # bill data only, then re-attach from the current votes index on the way
    # out (fresh records never carry it — Congress.gov isn't the vote source).
    merged, stats = merge_records(
        strip_vote_refs(existing.get("bills", [])), fresh_records
    )
    attach_vote_refs(merged, load_vote_refs(congress))
    final = save_manifest(congress, {"bills": merged})
    rebuild_index()

    print(
        f"merge: +{stats.added} added, ~{stats.updated} updated, "
        f"={stats.unchanged} unchanged (manifest now {len(final['bills'])} bills)"
    )
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: pipeline/shared/src/commonMain/kotlin/com/informedcitizen/pipeline/fetch/FetchBills.kt
package com.informedcitizen.pipeline.fetch

import com.informedcitizen.pipeline.ErrorCollector
import com.informedcitizen.pipeline.http.CongressClient
import com.informedcitizen.pipeline.model.Bill
import com.informedcitizen.pipeline.model.BillsManifest
import kotlinx.datetime.Instant
import kotlinx.serialization.json.JsonObject

/** Recent-window cutoff in days. Mirrors Python `fetch_bills.RECENT_DAYS = 60`. */
const val RECENT_DAYS: Int = 60

/** Number of `no_outcome_match` examples to surface in run logs. Mirrors Python `SAMPLE_REJECTIONS = 8`. */
const val SAMPLE_REJECTIONS: Int = 8

/**
 * Sentinel for [fetchBills]'s `maxNew`: enrich every kept summary with
 * no per-run cap. The CLI/CI default — a server has the full 5,000
 * requests/hour budget and re-enriches bills to pick up in-window
 * updates. The BYOK path (backlog #43) passes a real cap instead so a
 * phone never blows one key's hourly budget in a single tick.
 */
const val NO_ENRICHMENT_CAP: Int = Int.MAX_VALUE

/** Summary of a single fetch-bills run, returned by [fetchBills] for CLI output. */
data class FetchBillsResult(
    val congress: Int,
    val cutoff: Instant,
    val evaluated: Int,
    val keptRecords: List<Bill>,
    val rejectionCounts: Map<String, Int>,
    val rejectionSamples: List<String>,
    val mergeStats: MergeStats,
    val finalManifest: BillsManifest,
    /**
     * Bills that matched the filter and are not yet in the manifest but
     * were left for a future run because the `maxNew` enrichment cap was
     * reached (backlog #43). Always 0 for an uncapped run; a nonzero
     * value means a capped (BYOK) run will top up on its next tick.
     */
    val newBillsDeferred: Int = 0,
)

/**
 * Full fetch-bills orchestrator. Mirrors the Python `fetch_bills.main`
 * flow:
 *
 *  1. Paginated list (filtered by `fromDateTime = cutoff`).
 *  2. Per-summary filter ([evaluateBill]); tally rejections.
 *  3. Parallel enrichment ([buildBillRecordsParallel]).
 *  4. Dedupe by id (rare; same bill can appear twice across pages
 *     when its updateDate shifts mid-walk).
 *  5. Sort by latest_action.date desc.
 *  6. Merge into the existing manifest ([mergeBillRecords]), with the
 *     derived `votes` field stripped from existing records first and
 *     re-attached from [votesStore]'s index after ([attachVoteRefs]).
 *  7. Persist via [manifestStore].
 *
 * [maxNew] caps how many *new* bills (those not already in the manifest)
 * are enriched this run — the expensive step, three Congress.gov GETs
 * each. The default [NO_ENRICHMENT_CAP] disables the cap and re-enriches
 * every kept summary, so the CLI/CI behaviour (and its Python parity) is
 * unchanged. A finite [maxNew] (the BYOK path, backlog #43) instead
 * skips summaries already in the manifest for free — the manifest is the
 * resume cursor, mirroring how [fetchVotes] skips a roll call already on
 * disk — and spends the budget only on bills not yet fetched, deferring
 * the rest to a later run ([FetchBillsResult.newBillsDeferred]).
 *
 * Note: this slice does NOT rewrite `congresses.json` — the CLI
 * wrapper ([com.informedcitizen.pipeline.cli.FetchBillsCommand])
 * does that as a separate step after this returns, mirroring
 * Python `fetch_bills.main`'s flow (`save_manifest` then
 * `rebuild_index`). Keeps this orchestrator focused on fetch + merge.
 */
suspend fun fetchBills(
    client: CongressClient,
    congress: Int,
    cutoff: Instant,
    nowIso: String,
    manifestStore: FileBillsManifestStore,
    errors: ErrorCollector,
    votesStore: FileVotesStore? = null,
    maxListPages: Int = LIST_PAGES_MAX,
    maxWorkers: Int = ENRICH_WORKERS,
    maxNew: Int = NO_ENRICHMENT_CAP,
): FetchBillsResult {
    // Phase 1: paginated list + filter.
    val rejectCounts = mutableMapOf<String, Int>()
    val rejectionSamples = mutableListOf<String>()
    val keptSummaries = mutableListOf<Pair<JsonObject, String>>()
    var totalEvaluated = 0

    for (summary in listRecentBills(client, congress, cutoff, maxListPages)) {
        totalEvaluated++
        when (val result = evaluateBill(summary, cutoff)) {
            is BillEvaluationResult.Kept -> keptSummaries += summary to result.outcome
            is BillEvaluationResult.Rejected -> {
                rejectCounts.bump(result.reason)
                if (result.reason == RejectionReasons.NO_OUTCOME_MATCH &&
                    rejectionSamples.size < SAMPLE_REJECTIONS
                ) {
                    val latestAction = summary.jsonObjectField("latestAction")
                    val actionText = latestAction?.stringField("text").orEmpty()
                    val ref = "${summary.stringField("type")}${summary.stringField("number")}"
                    rejectionSamples += "$ref: ${actionText.take(140)}"
                }
            }
        }
    }

    // Load the existing manifest once: it is both the merge base and —
    // for a capped run — the resume cursor. Its `votes` field is derived
    // data, stripped here so the eventual equality merge compares bill
    // data only (re-attached on the way out).
    val existing = stripVoteRefs(manifestStore.load(congress)?.bills.orEmpty())

    // Choose which kept summaries to enrich. Uncapped: all of them
    // (re-enriching in-window updates, unchanged CLI/CI parity). Capped
    // (BYOK #43): only bills not already in the manifest, newest-updated
    // first (listRecentBills sorts updateDate desc), up to [maxNew]; the
    // rest are deferred to a future tick.
    val toEnrich: List<Pair<JsonObject, String>>
    val newBillsDeferred: Int
    if (maxNew == NO_ENRICHMENT_CAP) {
        toEnrich = keptSummaries
        newBillsDeferred = 0
    } else {
        val existingIds = existing.mapTo(mutableSetOf()) { it.id }
        val newSummaries = keptSummaries.filter { (summary, _) ->
            billIdForSummary(summary, congress) !in existingIds
        }
        toEnrich = newSummaries.take(maxNew.coerceAtLeast(0))
        newBillsDeferred = newSummaries.size - toEnrich.size
    }

    // Phase 2: parallel enrichment.
    val (freshRecords, buildFailures) =
        buildBillRecordsParallel(client, congress, toEnrich, errors, maxWorkers)
    if (buildFailures > 0) {
        rejectCounts.bump(RejectionReasons.BUILD_ERROR, buildFailures)
    }

    // Dedupe by id (rare cross-page repeats), then sort desc by
    // latest_action.date with id as tiebreaker — same ordering as
    // mergeBillRecords, so the "+ id" log lines are deterministic.
    val seenIds = mutableSetOf<String>()
    val deduped = mutableListOf<Bill>()
    for (rec in freshRecords) {
        if (rec.id in seenIds) {
            rejectCounts.bump(RejectionReasons.DUPLICATE)
            continue
        }
        seenIds += rec.id
        deduped += rec
    }
    val sorted = deduped.sortedWith(
        compareByDescending<Bill> { it.latestAction.date }.thenBy { it.id }
    )

    // Merge into the existing manifest and persist. `votes` is derived
    // data: `existing` was already stripped above, and fresh records
    // never carry it (Congress.gov isn't the vote source), so equality
    // compares bill data only; the derived `votes` is re-attached from
    // the current votes index on the way out. Mirrors Python
    // `fetch_bills.main`'s strip/attach pair.
    val (merged, mergeStats) = mergeBillRecords(existing, sorted)
    val enriched = attachVoteRefs(merged, votesStore?.loadVoteRefs(congress).orEmpty())
    val finalManifest = manifestStore.save(congress, enriched, nowIso)

    return FetchBillsResult(
        congress = congress,
        cutoff = cutoff,
        evaluated = totalEvaluated,
        keptRecords = sorted,
        rejectionCounts = rejectCounts.toMap(),
        rejectionSamples = rejectionSamples.toList(),
        mergeStats = mergeStats,
        finalManifest = finalManifest,
        newBillsDeferred = newBillsDeferred,
    )
}

/**
 * The [Bill.id] a list-endpoint summary will enrich into —
 * `"${type}${number}-${congress}"` with a lowercased type, matching
 * [buildBillRecord]. Lets a capped run (backlog #43) recognise a summary
 * already in the manifest without paying its three-GET enrichment cost.
 */
private fun billIdForSummary(summary: JsonObject, congress: Int): String {
    val type = (summary.stringField("type") ?: "").lowercase()
    val number = summary.stringField("number") ?: ""
    return "$type$number-$congress"
}

// `MutableMap.merge` is JVM-only — Kotlin Native (iOS targets) lacks it.
// Hand-rolled increment keeps commonMain portable.
private fun MutableMap<String, Int>.bump(key: String, by: Int = 1) {
    this[key] = (this[key] ?: 0) + by
}
